README.rst
**********************
net.cdaniels.toolchest
**********************

**ATTENTION**: by using this software, you agree to abide by the license
provided in each software module. If a software module does not have a license
explicitly provided, it falls under the `project license <LICENSE>`_.

.. contents::


Overview
========

`CHANGELOG <CHANGELOG>`_

`ROADMAP <ROADMAP>`_

This is a collection of tools and utilities that I and other's have written. This
project is a combination of two parts. The first is the ``toolchest`` itself, a
simple package manager for installing, uninstalling, and managing tools. The
second component is a collection of *packages*, which are mostly scripts and 
small tools which are too small to warrant being standalone software packages. 

Additionally, there are a number of *ports*, which are software tools written by
others, and are useful in cases where one is using a system with no package
manager, or a package manager which is unavailable for some reason. Ports are
internally identical to packages and are managed and interacted with in the same
way.

The toolchest attempts to fulfill a few key goals, which are elaborated on in
greater detail later; in short, they are:

* portability - the toolchest should ideally run on any POSIX-ish system that
  supports Bourne shell or 100% compatible.

* independance - the toolchest should not require any dependencies not installed
  out of the box on a supported system

* self-contained - the toolchest does not require root access, and all files
  system operations happen inside a single prefix (and ``/tmp``); deleting the
  toolchest completely removes all components from the system

Note that individual packages do not necessarily follow all of the above, with
the exception of the last goal, which is rigorously observed in all packages
I have written. 

Quick Installation Guide
------------------------

1. ``cd ~``
2. ``git clone https://bitbucket.org/charlesdaniels/toolchest``
3. ``mv toolchest .net.cdaniels.toolchest``
4. ``cd .net.cdaniels.toolchest``
5. ``./bin/toolchest setup``
6. ``./bin/toolchest-genpath >> ~/.bashrc``
7. ``./bin/toolchest-genpath >> ~/.bash_profile``
8. ``./bin/toolchest-genpath >> ~/.profile``
   
You may need to open a new shell session before the toolchest tools appear in 
``$PATH``. 

For documentation on using the ``toolchest`` command, run ``toolchest-doc
toolchest``.

For more information, see the `User Manual <./doc/manual.rst>`_ .

Goals
-----

The original purpose of this project was to ensure that systems specific
artifacts (eg. compiled binaries) are kept fully separated from the platform-
independent source code. Before switching to this method, I kept everything in
~/bin and synced this via git. This works pretty well for shell scripts, but
anything compiled does not play well with that approach, and even scripts that
have special dependencies could not be guaranteed to work without manually
validating that their dependencies were present. This is still a primary concern
to the toolchest, although further goals have been added over time.

The toolchest, as noted in the overview, aims to be highly portable. While not
all packages can always be portable to all systems, the ``toolchest`` itself and
the core utilities that ship with it will run with only Bourne shell (usually sh
or BASH) and common UNIX utilities. A system which has a C compiler, a python 3
interpreter, and a Perl 5 interpreter will run almost all packages supplied in
the default install.

The toolchest aims to run without any dependencies on supported systems. It will
work correctly (although all packages will not) out of the box on a stock
install of Ubuntu 16 or macOS 10.12, as well as many other operating systems.
Full functionality can be attained with minimal dependencies which most
developers and system administrates will already have installed. Further, almost
all packages have been configured with dependency validation which occurs before
installation, to avoid situations where a package installs but is broken after
the fact.

Last, the toolchest aims to be fully self-contained. Everything, including
packages, lives under the installation directory. This means installing the
toolchest is as simple as copying the install folder somewhere convenient, and
installing is simply the act of deleting the installation folder. 

Design
------

**NOTE**: except as noted, relative paths are assumed to be relative to the
toolchest installation directory.

In this repository, we keep a set of core utilities (usually referred to as
coreutils, or uniquely net.cdaniels.toolchest.coreutils) in the bin/ folder.
This lives directly under the toolchest install directory. These utilities
include the toolchest management utility itself, as well as it's dependencies,
and various cross-platform scripts which are used commonly by packages.

Additionally, a set of core libraries is provided in the lib/ folder which is
also stored directly under the toolchest install directory. These contain very
commonly used tools which are not useful on their own. This includes the
reference package installer and uninstaller scripts, among other things.

Packages are kept in the packages/ directory. Packages install using either a
makefile or a set of scripts (install.sh and uninstall.sh). Package artifacts
are symlinked into local/lib or local/bin as appropriate (local/ is directly
under the toolchest installation directory). 

Packages may be managed using ``toolchest package``. 

Beginning with 0.0.5-ALPHA, dependency resolution among toolchest packages
is be supported as part of the toolchest management utility.

As of the 1.X.X release series, packages are further organized into
*repositories*. See the `Packaging Guide <doc/packaging.rst>`_ for further
information.

Packages are not to store any configuration files in local/bin or local/lib.
However, packages may use local/config for storing configuration if they need
it.


Rationale
---------

In many ways, this project is re-inventing the wheel. There are plenty of
excellent package manager out there. However, I felt that this approach was
useful enough to warrant a new project, for a number of reasons:

* the toolchest (should) run on any POSIX system 
* the toolchest does not need to be run as root
* the toolchest has very few dependencies, just a POSIX shell and some common
  utilities
* the toolchest (as of 0.0.5) supports running from any location - this makes
  it suitable for using as a runtime.



Style and Conventions
=====================

Documentation about specific packages and scripts is kept in the relevant source
files. The motto for documentation in this project is *code is prose* - all
source code files should contain all of their documentation in-line. In some
cases, it is necessary to store overarching documentation that is not module
specific - this is kept in the `doc <./doc>`_ folder.

We use 80 character line width for all files, although this is not strict. Some
shell scripts need to exceed this here and there for long conditional
statements. Ideally < 1% of total lines of code exceed 80 characters. This makes
the source code convenient to work with in a variety of editors and on a variety
of screen sizes. Documentation files are also hard-wrapped to 80 characters for
the same reason, although neither reStructuredText nor Markdown require this.

Things which are used in 3 or more packages should probably be moved into
coreutils (lib or bin), unless they have many dependencies or require
compilation, in which case they should be moved into their own package. This
make re-use easier. 

Libraries that cannot be executed on their own (eg. that just contain function
declarations) should end in ``.lib`` and should not be marked executable. It has
recently come to my attention that some compilers also emit ``.lib`` files as a
form of binary object file - the ``.lib`` files used in the toolchest should not
be confused. All toolchest ``.lib`` files are shell scripts which can be
sourced. 

Libraries that contain functions should not executed anything when
sourced beyond said function declarations.

From the point of view of the rest of the toolchest, the coreutils libraries
and binaries should be considered non-mutable. Outside of toolchest updates,
they should not change in contents or functionality, and their behavior is
not configurable.

Gotchas
-------

There are a few gotchas that I have identified which can affect cross platform
compatibility. This information is mostly of interest to individual wishing to
develop toolchest packages.

* On most systems in sh, ``/foo/bar/somefile/..`` is a valid path that points
  to ``/foo/bar``. This does NOT work on FreeBSD 11's sh however. 

* The convention for copying directory structures recursively differs between
  various UNIXes. Rather than writing an ugly switch statement, ``( cd
  /path/to/src ; tar cf - . ) | ( cd /path/to/dest ; tar xf - )`` will do the
  same thing, albeit somewhat awkwardly. Note that this functionality is now a
  part of the coreutils and is named `mirror-directory <./bin/mirror-
  directory>`_ .(`source <http://superuser.com/a/138604>`_)


Further Documentation
=====================

* `acquiring the toolchest installation directory <doc/paths.rst>`_
* `user manual <doc/manual.rst>`_
* `packaging guide <doc/packaging.rst>`_
* `versioning scheme <doc/version.rst>`_
* `toolchest management tool <bin/toolchest>`_
* `system requirements & supported OSes <doc/support.rst>`_
* `reference package installer <lib/generic-install.sh>`_
* `reference package uninstaller <lib/generic-uninstall.sh>`_
* `example with sh installer <packages/example>`_
* `example with make installer <packages/helloworld>`_


CHANGELOG
0.0.2-ALPHA:

  2016-12-12 - Charles Daniels:

    * Fixed an issue affecting FreeBSD11 (and probably other BSD systems)
      where $0/../.. would error because $0 was a file, not a directory.
      This expansion is legal in many sh implementations (including macOS
      10.12), but failed on FreeBSD 11.

      This was fixed by removing the `toolchest` from the end of `realpath
      $0` using sed and working relative to that, as the parent of a file
      should always be a directory.

0.0.3-ALPHA:

  2016-12-12 - Charles Daniels:

    * Installer now uses tar in a rather novel way to copy directories while
      preserving symlinks. This fixes an issue where the installer would
      copy links to generic-installer and generic-uninstaller as if they
      were files with the actual contents of the relevant libraries.

      This method was chosen instead of cp --preserve=links because BSD-
      style systems (tested: macOS 10.12, FreeBSD 11) do not support this
      syntax for cp.

    * Added additional progress messages to installer

    * Added random-uuid to universal-tools package

  2016-12-13 - Charles Daniels:

    * Fixed a bug in acquire-toolchest-dirs which caused it to incorrectly
      identify NET_CDANIELS_TOOLCHEST_DIR

    * Added check-command-exists library to coreutils libraries

    * universal-tools now has a preinstall script to validate it's dependencies

    * archiveutils now has a preinstall script to validate it's dependencies

  2016-12-16 - Charles Daniels

    * renamed the mac-shortcuts package to macosutils to better reflect purpose

    * renamed cleardns to macosutils-cleardns

    * added an alternate method of DNS cache flushing to macosutils-cleardns

    * macosutils-cleardns no longer kills mDNSResponder as root - this may
      reduce it's effectiveness when running as non-root users

    * renamed unquarintine to macosutils-unquarantine

    * added macosutils-fixwifi script to troubleshoot common end-user network
      problems

  2016-12-16 - Charles Daniels

    * added package to install m-cli

0.0.4-ALPHA:

  2016-12-19 - Charles Daniels:

    * Wrote the carbonite 0.0.1 file format specification

  2016-12-20 - Charles Daniels:

    carbonite:

      - argument parsing

      - logging (via loggingutils)

    * added the net.cdaniels.python package with lib/loggingutils.py package

    * added `dir` directive to `toolchest`

  2016-12-21 - Charles Daniels:

    net.cdaniels.toolchest.coreutils:

      sensible-python3

        * add support for Python 3.6

        * add support for Python 3.7

    carbonite:

      * payload and include validation

      * checksum generation

      * archive generation

      * can produce valid output via --archive_only

      * installer (via Makefile)

      * sh wrapper

      * payload extract-execute script

      * carbonite standalone executable generation

0.0.5-ALPHA

  2016-12-21 - Charles Daniels:

    * implemented function to check if packages are installed, and to mark
      them installed or uninstalled.

      * still need to make the package install/uninstall directives interface
        with the new functions.

      * implemented "toolchest package mark-installed"

      * implemented "toolchest package mark-uninstalled"

  2016-12-22 - Charles Daniels:

    * add iproute2mac package

    * toolchest management utility now correctly interacts with installation
      status. For example, it will refuse to install a package which is
      already installed.

    * implemented "toolchest package list-installed"

    * "toolchest package mark-[un]installed" now refuses to work on
      nonexistent packages

    pdfutils:

      * add support for PDFPen

      * fixed a bug where paths containing spaces would result in invoke-*-ocr
        being called on a garbage path.

      * installer now warns if no supported OCR software is detected

    * toolchest management utility now has internal functions to perform
      unconditional package installations and uninstallations.

    * toolchest now has an "internal" directive for internal use commands

    * implemented internal use commands:

      * internal package unconditional-install

      * internal package unconditional-uninstall

      * internal package mark-installed

      * internal package mark-uninstalled

    * removed package mark-installed and mark-uninstalled, as these were
      intended to be internal use only from the getgo (the same functionality
      is now provided by the internal commands above).

    * implemented `toolchest refresh` to refresh toolchest installation,
      preserving only configuration and installed package listing.

    * toolchest package list-installed now displays an error message when the
      package list has not yet been initialized.

  2016-12-24 - Charles Daniels:

    * added package for tcc c compiler

  2016-12-26 - Charles Daniels:

    * added user's manual

    * added packaging guide

  2016-12-30 - Charles Daniels:

    * added normalize-path implementation in...

      * Python

      * pure SH

      * BASH + readlink

    * added normalize-path to net.cdaniels.coreutils as a binary in $PATH, to
      provide cross-platform path normalization - it should be used to replace
      realpath in all scripts from here out.

  2016-12-31 - Charles Daniels:

    * acquire-toolchest-dirs now backed by `toolchest dir`

    * documented conventions for acquiring toolchest install dir

    * added directives to ``acquire-toolchest-dirs``, implemented...

      * export - produces legacy functionality

      * dironly - identical behaviour to ``toolchest dir``, intended as a
        stable API to provide this functionality for scripts

    * replaced normalize-path with Michael Kropat's sh implementation from
      GitHub. normalize-path is still left as a coreutils binary for
      convienience, but it simply wraps the new lib/realpath.lib

  2017-01-01 - Charles Daniels:


lib/generic-uninstall.sh
#!/bin/sh
########10########20########30## DOCUMENTATION #50########60########70########80
#
#  OVERVIEW
#  ========
#  
#  Uninstalls things installed by generic-install.sh. 
#  
#  Note that this script depends on the current working directory being the
#  package root. If it is not, things will definitely break in unpredictable
#  ways. To this end, you should probably not run this by hand. 
#    
########10########20########30##### LICENSE ####50########60########70########80
#  Copyright (c) 2016, Charles Daniels
#  All rights reserved.
# 
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:
# 
#  1. Redistributions of source code must retain the above copyright notice,
#     this list of conditions and the following disclaimer.
#  
#  2. Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
# 
#  3. Neither the name of the copyright holder nor the names of its
#     contributors may be used to endorse or promote products derived from
#     this software without specific prior written permission.
# 
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#  POSSIBILITY OF SUCH DAMAGE.
#  
########10########20########30########40########50########60########70########80

# make sure the caller knows what toolchest it's working with
if ! (: "${NET_CDANIELS_TOOLCHEST_DIR?}") 2>/dev/null; then
  echo "ERROR 7: NET_CDANIELS_TOOLCHEST_DIR is not defined, installation failed"
  exit 1
fi

if ! (: "${PACKAGE_NAME?}") 2>/dev/null; then
  echo "ERROR 68: PACKAGE_NAME is not defined, installation failed"
  exit 1
fi

if ! (: "${PACKAGE_PATH?}") 2>/dev/null; then
  echo "ERROR 7: PACKAGE_PATH is not defined, installation failed"
  exit 1
fi

# setup some convenient shortcuts
DEST_DIR="$NET_CDANIELS_TOOLCHEST_DIR/local/bin"
SRC_DIR="$PACKAGE_PATH/bin"

# sanity check - make sure we are installing from the toolchest the caller
# thinks we are installing from
if [ `realpath ./bin` != "$SRC_DIR" ] ; then
  echo "ERROR 19: package source directory mismatch. This is really bad. You"
  echo "          seem to have NET_CDANIELS_TOOLCHEST_DIR pointing somewhere"
  echo "          other than your toolchest installation"
  echo "DEBUG: `realpath ./bin` does not match $SRC_DIR"
  exit 1
fi

# this gives us just the filename (eg. file1 file2 file 3 as opposed to
# bin/file1 bin/file2 bin/file3 and so on)
cd ./bin/

# loop over everything in bin/ and setup the symlinks
for f in * ; do
  SRC_FILE="$SRC_DIR/$f"
  DEST_FILE="$DEST_DIR/$f"
  printf "INFO: (bin) unlinking $f... "

  # make sure the thing we are trying to uninstall exists and is a link
  if [ -L "$DEST_FILE" ] ; then
    rm "$DEST_FILE"
  
    if [ $? -ne 0 ] ; then
      echo "FAIL"
      echo "ERROR 42: failed to remove link"
      exit 1
    fi
  elif [ -e "$DEST_FILE" ] ; then
    echo "WARN"
    echo "WARNING 46: $DEST_FILE exists, but is not a symlink, skipping it"
  fi

  echo "DONE"
done

cd ..
if [ -d lib ] ; then
  # for convenience, we go relative to the already calculated directories
  LIB_DIR="$DEST_DIR/../lib/$PACKAGE_NAME"
  
  if [ ! -d "$LIB_DIR" ] ; then
    echo "INFO: (lib) no library directory present"
    break
  fi

  cd lib 
  for f in * ; do
    printf "INFO: (lib) unlinking $f... "

    if [ -L "$LIB_DIR/$f" ] ; then
      rm "$LIB_DIR/$f"
    elif [ -e "$LIB_DIR/$f" ] ; then
      echo "WARN"
      echo "WARNING 104: $LIB_DIR/$f exists but is not a symlink!"
    fi
    echo "DONE"
  done
  cd ..

  printf "INFO: (lib) removing library directory for $PACKAGE_NAME..."
  rmdir "$LIB_DIR"
  echo "DONE"
fi


lib/normalize-path/normalize-path.py
########10########20########30## DOCUMENTATION #50########60########70########80
#
#  OVERVIEW
#  ========
#
#  Provides a very basic implementation of realpath that will run in python. 
#  
#  Should run in Python 2 or 3 (tested in 2.7 and 3.6).
#  
#  USAGE
#  =====
#  
#  $1 . . . path to normalize
#  
#  RETURNS
#  =======
#  
#  Outputs the normalized path to stdout. 
#
########10########20########30##### LICENSE ####50########60########70########80
#  Copyright (c) 2016, Charles Daniels
#  All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:
#
#  1. Redistributions of source code must retain the above copyright notice,
#     this list of conditions and the following disclaimer.
#
#  2. Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
#
#  3. Neither the name of the copyright holder nor the names of its
#     contributors may be used to endorse or promote products derived from
#     this software without specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#  POSSIBILITY OF SUCH DAMAGE.
#
########10########20########30## CONFIGURATION #50########60########70########80


import os
import sys

if len(sys.argv) != 2: 
  print("ERROR 55: no path specified")
  exit(1)

try:
  print(os.path.abspath(sys.argv[1]))
except Exception as e:
  print("ERROR 61: exception while computing absolute path of {}"
        .format(sys.argv[1]))
  print("INFO: exception details follow: \n{}".format(e))
  exit(1)



lib/normalize-path/normalize-path.sh
########10########20########30## DOCUMENTATION #50########60########70########80
#
#  OVERVIEW
#  ========
#
#  Realpath implementation in pure sh. This implementation should only be used
#  on systems without readlink and BASH, and where realpath is not present.
#  
#  Note that this may produce "dirty" paths, such as /foo/bar/./, although
#  these will be correct, and should resolve correctly by things that parse
#  paths in a POSIX compliant way.
#  
#  USAGE
#  =====
#  
#  $1 . . . path to normalize
#  
#  RETURNS
#  =======
#  
#  Outputs the normalized path to stdout. 
#
########10########20########30##### LICENSE ####50########60########70########80
#
#  This script sourced from StackOverflow user Sildoreth
#  
#  It has been slightly modified to fit net.cdaniels.toolchest conventions
#  
#  Original URL: 
#
#  http://unix.stackexchange.com/a/204381
#
########10########20########30## CONFIGURATION #50########60########70########80


if [ $# -eq 0 ] || [ $# -gt 2 ]; then
  printf "Usage: $0 path [working-directory]\n" >&2
  exit 1
fi
cantGoUp=

path=$1
if [ $# -gt 1 ]; then
  cd "$2"
fi
cwd=`pwd -P` #Use -P option to account for directories that are actually symlinks

#Handle non-relative paths, which don't need resolution
if echo "$path" | grep '^/' > /dev/null ; then
  printf '%s\n' "$path"
  exit 0
fi

#Resolve for each occurrence of ".." at the beginning of the given path.
#For performance, don't worry about ".." in the middle of the path.
while true
do
  case "$path" in
    ..*)
      if [ "$cwd" = '/' ]; then
        printf 'Invalid relative path\n' >&2
        exit 1
      fi
      if [ "$path" = '..' ]; then
        path=
      else
        path=`echo "$path" | sed 's;^\.\./;;'`
      fi
      cwd=`dirname $cwd`
      ;;
    *)
      break
      ;;
  esac
done

cwd=`echo "$cwd" | sed 's;/$;;'`
if [ -z "$path" ]; then
  if [ -z "$cwd" ]; then
    cwd='/'
  fi
  printf '%s\n' "$cwd"
else
  printf '%s/%s\n' "$cwd" "$path"
fi

lib/normalize-path/normalize-path.bash
########10########20########30## DOCUMENTATION #50########60########70########80
#
#  OVERVIEW
#  ========
#
#  Realpath implementation in BASH (NOT sh) by David Raistrick. Depends on
#  BASH and readlink. For systems without realpath, but with bash and
#  readlink, this is the preferred implementation.
#  
#  This appears to also work with zsh. 
#  
#  USAGE
#  =====
#  
#  $1 . . . path to normalize
#  
#  RETURNS
#  =======
#  
#  Outputs the normalized path to stdout. 
#
########10########20########30##### LICENSE ####50########60########70########80
#
#  This script sourced from David Raistrick's GitHub page on 2016-12-30
#  
#  It has been slightly modified to fit net.cdaniels.toolchest conventions. 
#  
#  All credit to David Raistrick
#  
#  Original URL: 
#
#  https://github.com/keen99/shell-functions/tree/master/resolve_path
#
########10########20########30## CONFIGURATION #50########60########70########80



#I'm bash only, please!
# usage:  resolve_path <a file or directory> 
# follows symlinks and relative paths, exits a full real path
#
owd="$PWD"
#echo "$FUNCNAME for $1" >&2
opath="$1"
npath=""
obase=$(basename "$opath")
odir=$(dirname "$opath")
if [[ -L "$opath" ]]
then
#it's a link.
#file or directory, we want to cd into it's dir
  cd $odir
#then extract where the link points.
  npath=$(readlink "$obase")
  #have to -L BEFORE we -f, because -f includes -L :(
  if [[ -L $npath ]]
   then
  #the link points to another symlink, so go follow that.
    resolve_path "$npath"
    #and finish out early, we're done.
    exit $?
    #done
  elif [[ -f $npath ]]
  #the link points to a file.
   then
    #get the dir for the new file
    nbase=$(basename $npath)
    npath=$(dirname $npath)
    cd "$npath"
    ndir=$(pwd -P)
    retval=0
    #done
  elif [[ -d $npath ]]
   then
  #the link points to a directory.
    cd "$npath"
    ndir=$(pwd -P)
    retval=0
    #done
  else
    echo "$FUNCNAME: ERROR: unknown condition inside link!!" >&2
    echo "opath [[ $opath ]]" >&2
    echo "npath [[ $npath ]]" >&2
    exit 1
  fi
else
  if ! [[ -e "$opath" ]]
   then
    echo "$FUNCNAME: $opath: No such file or directory" >&2
    exit 1
    #and break early
  elif [[ -d "$opath" ]]
   then 
    cd "$opath"
    ndir=$(pwd -P)
    retval=0
    #done
  elif [[ -f "$opath" ]]
   then
    cd $odir
    ndir=$(pwd -P)
    nbase=$(basename "$opath")
    retval=0
    #done
  else
    echo "$FUNCNAME: ERROR: unknown condition outside link!!" >&2
    echo "opath [[ $opath ]]" >&2
    exit 1
  fi
fi
#now assemble our output
echo -n "$ndir"
if [[ "x${nbase:=}" != "x" ]]
 then
  echo "/$nbase"
else 
  echo
fi
#now exit to where we were
cd "$owd"
exit $retval

lib/generic-install.sh
#!/bin/sh
########10########20########30## DOCUMENTATION #50########60########70########80
#
#  OVERVIEW
#  ========
#  
#  Generic package installer script. This is most useful for simple scripts
#  that don't require complex compilation or dependency checking. This script
#  can simply be symlinked into the package directory (don't forget to call
#  the destination install.sh).
#   
#  This script assumes that all the things you want to be in $PATH are in bin/
#  under the package root, with the names you want them to have.
#  
#  This script will also check to see if lib/ exists under the package root,
#  and if so, will symlink everything in it into local/lib/$PACKAGNAME
#  
#  Note that this script depends on the current working directory being the
#  package root. If it is not, things will definitely break in unpredictable
#  ways. To this end, you should probably not run this by hand. 
#  
#  If the file `preinstall.sh` is present in the top level of the package
#  directory, generic-install will call it before performing the installation.
#  This is particularly useful for checking that dependencies are installed.
#  If the preinstall script exits with a nonzero exit code, generic-install
#  will also exit with a nonzero exit code, without performing the
#  installation.
#    
########10########20########30##### LICENSE ####50########60########70########80
#  Copyright (c) 2016, Charles Daniels
#  All rights reserved.
# 
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:
# 
#  1. Redistributions of source code must retain the above copyright notice,
#     this list of conditions and the following disclaimer.
#  
#  2. Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
# 
#  3. Neither the name of the copyright holder nor the names of its
#     contributors may be used to endorse or promote products derived from
#     this software without specific prior written permission.
# 
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#  POSSIBILITY OF SUCH DAMAGE.
#  
########10########20########30########40########50########60########70########80

# make sure the caller knows what toolchest it's working with
if ! (: "${NET_CDANIELS_TOOLCHEST_DIR?}") 2>/dev/null; then
  echo "ERROR 7: NET_CDANIELS_TOOLCHEST_DIR is not defined, installation failed"
  exit 1
fi

if ! (: "${PACKAGE_NAME?}") 2>/dev/null; then
  echo "ERROR 68: PACKAGE_NAME is not defined, installation failed"
  exit 1
fi

if ! (: "${PACKAGE_PATH?}") 2>/dev/null; then
  echo "ERROR 7: PACKAGE_PATH is not defined, installation failed"
  exit 1
fi

# make sure realpath is available
. "$NET_CDANIELS_TOOLCHEST_DIR/lib/realpath.lib"

# run preinstall script, if present
if [ -e ./preinstall.sh ] ; then
  ./preinstall.sh 
  if [ $? -ne 0 ] ; then
    echo "ERROR 61: preinstall script failed"
    exit 1
  fi
fi

# setup some convenient shortcuts
DEST_DIR="$NET_CDANIELS_TOOLCHEST_DIR/local/bin"
SRC_DIR="$PACKAGE_PATH/bin"

# sanity check - make sure we are installing from the toolchest the caller
# thinks we are installing from
if [ `realpath ./bin` != "$SRC_DIR" ] ; then
  echo "ERROR 19: package source directory mismatch. This is really bad. You"
  echo "          seem to have NET_CDANIELS_TOOLCHEST_DIR pointing somewhere"
  echo "          other than your toolchest installation"
  echo "DEBUG: `realpath ./bin` does not match $SRC_DIR"
  exit 1
fi

# this gives us just the filename (eg. file1 file2 file 3 as opposed to
# bin/file1 bin/file2 bin/file3 and so on)
cd ./bin/

# loop over everything in bin/ and setup the symlinks
for f in * ; do
  SRC_FILE="$SRC_DIR/$f"
  DEST_FILE="$DEST_DIR/$f"
  printf "INFO: (bin) linking $f... "

  # make sure the destination does not already exist
  if [ -e "$DEST_FILE" ] ; then
    echo "FAIL"
    echo "ERROR 34: destination file $DEST_FILE already exists"
    exit 1
  fi

  ln -s "$SRC_FILE" "$DEST_FILE"

  # make sur ethe link operation woked
  if [ $? -ne 0 ] ; then
    echo "FAIL"
    echo "ERROR 42: failed to create link"
    exit 1
  fi

  echo "DONE"
done

# install any libraries if they exists
cd ..
if [ -d lib ] ; then
  # for convenience, we go relative to the already calculated directories
  LIB_DIR="$DEST_DIR/../lib/$PACKAGE_NAME"
  if [ -d "$LIB_DIR" ] ; then
    # lib dir can't already exist
    echo "ERROR 123: lib dir \"$LIB_DIR\" already exists!"
    exit 1
  fi
  mkdir "$LIB_DIR"
  cd lib 
  for f in * ; do
    printf "INFO: (lib) linking $f... "

    if [ -e "$LIB_DIR/$f" ] ; then
      # library in question can't already exist
      echo "FAIL"
      echo "ERROR 132: destination file "$LIB_DIR/$f" already exists!"
      exit 1
    fi

    # do the link operation
    ln -s "$SRC_DIR/../lib/$f" "$LIB_DIR/$f"

    if [ $? -ne 0 ] ; then
      # make sure it worked
      echo "FAIL"
      echo "ERROR 138: failed to create link from \"$SRC_DIR/../lib/$f\" to \"$LIB_DIR/$f\""
      exit 1
    fi
    echo "DONE"
  done
fi
